packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.tsx

```diff
@@ -0,0 +1,257 @@
+      if (registeredInput.current) {
+        console.error(
+          [
+            'Material-UI: There are multiple `InputBase` components inside a FormControl.',
+            'This creates visual inconsistencies, only use one `InputBase`.',
+          ].join('\n'),
+        );
+      }
+
+      registeredInput.current = true;
+      return () => {
+        registeredInput.current = false;
+      };
+    };
+  }
+
+  const onFilled = React.useCallback(() => {
+    setFilled(true);
+  }, []);
+
+  const onEmpty = React.useCallback(() => {
+    setFilled(false);
+  }, []);
+
+  const childContext: FormControlState = {
+    adornedStart,
+    setAdornedStart,
+    disabled,
+    error,
+    filled,
+    focused,
+    hiddenLabel,
+    onBlur: () => {
+      setFocused(false);
+    },
+    onEmpty,
+    onFilled,
+    onFocus: () => {
+      setFocused(true);
+    },
+    registerEffect,
+    required,
+    ...extraContextProperties,
+  };
+
+  const Root = component ?? components.Root ?? 'div';
+  const rootProps = appendStyleProps(Root, { ...other, ...componentsProps.root }, styleProps);
+
+  return (
+    <FormControlContext.Provider value={childContext}>
+      <Root
+        ref={ref}
+        {...rootProps}
+        className={clsx(
+          classes.root,
+          className,
+          rootProps?.className,
+          disabled && classes.disabled,
+        )}
+      >
+        {children}
+      </Root>
+    </FormControlContext.Provider>
+  );
+});
+
+FormControl.propTypes /* remove-proptypes */ = {
+  // ----------------------------- Warning --------------------------------
+  // | These PropTypes are generated from the TypeScript type definitions |
+  // |     To update them edit TypeScript types and run "yarn proptypes"  |
+  // ----------------------------------------------------------------------
+  /**
+   * The content of the component.
+   */
+  children: PropTypes.node,
+  /**
+   * Class name applied to the root element.
+   */
+  className: PropTypes.string,
+  /**
+   * The component used for the Root slot.
+   * Either a string to use a HTML element or a component.
+   * This is equivalent to `components.Root`. If both are provided, the `component` is used.
+   */
+  component: PropTypes.elementType,
+  /**
+   * The components used for each slot inside the FormControl.
+   * Either a string to use a HTML element or a component.
+   * @default {}
+   */
+  components: PropTypes.shape({
+    Root: PropTypes.elementType,
+  }),
+  /**
+   * @ignore
+   */
+  componentsProps: PropTypes.object,
+  /**
+   * If `true`, the label, input and helper text should be displayed in a disabled state.
+   * @default false
+   */
+  disabled: PropTypes.bool,
+  /**
+   * If `true`, the label is displayed in an error state.
+   * @default false
+   */
+  error: PropTypes.bool,
+  /**
+   * Extra properties to be placed on the FormControlContext.
+   * @default {}
+   */
+  extraContextProperties: PropTypes.object,
```

Comment:
Can be implemented via separate context. Merging context is not ok at the moment. Separate contexts don't have the same problems as merged context values.

Let's not add features before we've defined what they're used for.

Reply:
OK, do you have something like this in mind?

```jsx
<FormControl>
  <FormControlContext.Provider value={thingsSpecificToMd}>
    <FormControlUnstyled>
      <FormControlUnstyledContext.Provider value={genericThings}>
        <!-- actual form controls -->
```

and then we'd have MD-version of `useFormControl` that would call both `useContext(FormControlUnstyledContext)` and `useContext(FormControlContext)` and return their results?

Reply:
+1 for not adding this via props. The main reason is that it would be confusing if it exists on some components, but not on others. The code that you've shared @michaldudak makes sense to me.

Reply:
> OK, do you have something like this in mind?
> 
> ```js
> <FormControl>
>   <FormControlContext.Provider value={thingsSpecificToMd}>
>     <FormControlUnstyled>
>       <FormControlUnstyledContext.Provider value={genericThings}>
>         <!-- actual form controls -->
> ```

Yep 👍🏻 Everything else would be confusing since you'd have to always be aware in which kind of `FormControlUnstyledContext` you are i.e. which `value` is provided.